nova_vm/src/engine/bytecode/vm.rs

```diff
@@ -1390,50 +1390,40 @@ impl Vm {
     ) -> JsResult<()> {
         let value = vm.stack.pop().unwrap();
 
+        // 8.6.2 Runtime Semantics: BindingInitialization
+        // BindingPattern : ObjectBindingPattern
+        // 1. Perform ? RequireObjectCoercible(value).
+        // NOTE: RequireObjectCoercible throws in the same cases as ToObject, and other operations
+        // later on (such as GetV) also perform ToObject, so we convert to an object early.
+        let value = to_object(agent, value)?;
+
         loop {
             let instr = executable.get_instruction(&mut vm.ip).unwrap();
-            if instr.kind == Instruction::BindingPatternBind {
-                // Shorthand pattern, ie. SingleNameBinding: const { b } = a;
-                let binding_id = vm.fetch_identifier(executable, instr.args[0].unwrap() as usize);
-                let lhs = resolve_binding(agent, binding_id, environment)?;
-                let v = get_v(agent, value, binding_id.into())?;
-                if environment.is_none() {
-                    put_value(agent, &lhs, v)?;
-                } else {
-                    initialize_referenced_binding(agent, lhs, v)?;
-                }
-                continue;
-            } else if instr.kind == Instruction::EvaluatePropertyAccessWithIdentifierKey {
-                let property_name_string =
-                    vm.fetch_identifier(executable, instr.args[0].unwrap() as usize);
-                let strict = agent
-                    .running_execution_context()
-                    .ecmascript_code
-                    .unwrap()
-                    .is_strict_mode;
-
-                let reference = Reference {
-                    base: Base::Value(value),
-                    referenced_name: property_name_string.into(),
-                    strict,
-                    this_value: None,
-                };
+            match instr.kind {
+                Instruction::BindingPatternBind | Instruction::BindingPatternBindNamed => {
+                    let binding_id =
+                        vm.fetch_identifier(executable, instr.args[0].unwrap() as usize);
+                    let property_key = if instr.kind == Instruction::BindingPatternBind {
+                        binding_id.into()
+                    } else {
+                        PropertyKey::from_value(
```

Comment:
**thought:** It's a bit sad to check the integerness of the value every time, instead of doing the conversion once at compile time and storing a String or Number based on the result, and then doing essentially `PropertyKey::from_value_unchecked` here.